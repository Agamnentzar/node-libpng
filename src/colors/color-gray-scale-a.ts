export type ColorGrayScaleA = [number, number] & { gray: number, a: number };

export function colorGrayScaleA(gray: number, a: number): ColorGrayScaleA {
    const color = [gray, a];
    Object.defineProperty(color, "gray", { get() { return this[0]; } });
    Object.defineProperty(color, "a", { get() { return this[1]; } });
    return color as ColorGrayScaleA;
}

export function isColorGrayScaleA(color: any): color is ColorGrayScaleA {
    if (typeof color !== "object" || !Array.isArray(color)) { return false; }
    if (color.length !== 2) { return false; }
    if (typeof (color as any).gray !== "number") { return false; }
    if (typeof (color as any).a !== "number") { return false; }
    return true;
}
