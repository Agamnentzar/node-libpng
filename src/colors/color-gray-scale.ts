export type ColorGrayScale = [number] & { gray: number };

export function colorGrayScale(gray: number): ColorGrayScale {
    const color = [gray];
    Object.defineProperty(color, "gray", { get() { return this[0]; } });
    return color as ColorGrayScale;
}

export function isColorGrayScale(color: any): color is ColorGrayScale {
    if (typeof color !== "object" || !Array.isArray(color)) { return false; }
    if (color.length !== 1) { return false; }
    if (typeof (color as any).gray !== "number") { return false; }
    return true;
}
