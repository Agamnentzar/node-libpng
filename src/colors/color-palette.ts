export type ColorPalette = [number] & { index: number };

export function colorPalette(index: number): ColorPalette {
    const color = [index];
    Object.defineProperty(color, "index", { get() { return this[0]; } });
    return color as ColorPalette;
}

export function isColorPalette(color: any): color is ColorPalette {
    if (typeof color !== "object" || !Array.isArray(color)) { return false; }
    if (color.length !== 1) { return false; }
    if (typeof (color as any).index !== "number") { return false; }
    return true;
}
