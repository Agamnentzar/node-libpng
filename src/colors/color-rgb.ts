export type ColorRGB = [number, number, number] & { r: number, g: number, b: number };

export function colorRGB(r: number, g: number, b: number): ColorRGB {
    const color = [r, g, b];
    Object.defineProperty(color, "r", { get() { return this[0]; } });
    Object.defineProperty(color, "g", { get() { return this[1]; } });
    Object.defineProperty(color, "b", { get() { return this[2]; } });
    return color as ColorRGB;
}

export function isColorRGB(color: any): color is ColorRGB {
    if (typeof color !== "object" || !Array.isArray(color)) { return false; }
    if (color.length !== 3) { return false; }
    if (typeof (color as any).r !== "number") { return false; }
    if (typeof (color as any).g !== "number") { return false; }
    if (typeof (color as any).b !== "number") { return false; }
    return true;
}
