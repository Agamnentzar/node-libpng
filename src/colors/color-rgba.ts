export type ColorRGBA = [number, number, number, number] & { r: number, g: number, b: number, a: number };

export function colorRGBA(r: number, g: number, b: number, a: number): ColorRGBA {
    const color = [r, g, b, a];
    Object.defineProperty(color, "r", { get() { return this[0]; } });
    Object.defineProperty(color, "g", { get() { return this[1]; } });
    Object.defineProperty(color, "b", { get() { return this[2]; } });
    Object.defineProperty(color, "a", { get() { return this[3]; } });
    return color as ColorRGBA;
}

export function isColorRGBA(color: any): color is ColorRGBA {
    if (typeof color !== "object" || !Array.isArray(color)) { return false; }
    if (color.length !== 4) { return false; }
    if (typeof (color as any).r !== "number") { return false; }
    if (typeof (color as any).g !== "number") { return false; }
    if (typeof (color as any).b !== "number") { return false; }
    if (typeof (color as any).a !== "number") { return false; }
    return true;
}
