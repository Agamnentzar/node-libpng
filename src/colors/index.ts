export * from "./gray-scale-alpha";
export * from "./gray-scale";
export * from "./palette";
export * from "./rgb";
export * from "./rgba";
