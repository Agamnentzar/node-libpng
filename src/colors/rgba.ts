/**
 * Represents a color of color type `ColorType.RGBA`.
 *
 * @see ColorType
 */
export type ColorRGBA = [number, number, number, number] & { r: number, g: number, b: number, a: number };

/**
 * Create a new color of type `number`.
 *
 * @param r The value for the `red` part of the color.
 * @param g The value for the `green` part of the color.
 * @param b The value for the `blue` part of the color.
 * @param a The value for the alpha channel part of the color.
 *
 * @return The color in rgba representation.
 */
export function colorRGBA(r: number, g: number, b: number, a: number): ColorRGBA {
    const color = [r, g, b, a];
    Object.defineProperty(color, "r", { get() { return this[0]; } });
    Object.defineProperty(color, "g", { get() { return this[1]; } });
    Object.defineProperty(color, "b", { get() { return this[2]; } });
    Object.defineProperty(color, "a", { get() { return this[3]; } });
    return color as ColorRGBA;
}

/**
 * Checks if the given parameter is a color of type `ColorRGBA`.
 *
 * @param color The input to check.
 *
 * @return `true` if `color` was of type `ColorRGBA` and `false` otherwise.
 */
export function isColorRGBA(color: any): color is ColorRGBA {
    if (typeof color !== "object" || !Array.isArray(color)) { return false; }
    if (color.length !== 4) { return false; }
    if (typeof (color as any).r !== "number") { return false; }
    if (typeof (color as any).g !== "number") { return false; }
    if (typeof (color as any).b !== "number") { return false; }
    if (typeof (color as any).a !== "number") { return false; }
    return true;
}
