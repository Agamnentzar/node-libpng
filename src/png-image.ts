import { encode, writePngFile, writePngFileSync, WritePngFileCallback } from "./encode";
import { __native_PngImage } from "./native";

/**
 * The color type from libpng represented as a string.
 * This describes how the actual image pixels are defined.
 */
export enum ColorType {
    /**
     * Libpng color type `PNG_COLOR_TYPE_GRAY` (0).
     */
    GRAY = "gray",
    /**
     * Libpng color type `PNG_COLOR_TYPE_RGB` (2).
     */
    RGB = "rgb",
    /**
     * Libpng color type `PNG_COLOR_TYPE_PALETTE` (3).
     */
    PALETTE = "palette",
    /**
     * Libpng color type `PNG_COLOR_TYPE_GRAY_ALPHA` (4).
     */
    GRAY_ALPHA = "gray-alpha",
    /**
     * Libpng color type `PNG_COLOR_TYPE_RGB_ALPHA ` (6).
     */
    RGBA = "rgba",
    /**
     * Color type parsing failed.
     */
    UNKNOWN = "unknown",
}

/**
 * The interlace type from libpng.
 */
export enum InterlaceType {
    /**
     * Libpng interlace type `PNG_INTERLACE_NONE` (0).
     */
    NONE = "none",
    /**
     * Libpng interlace type `PNG_INTERLACE_ADAM7` (1).
     */
    ADAM7 = "adam7",
    /**
     * Interlace type parsing failed.
     */
    UNKNOWN = "unknown",
}

export class PngImage {
    private nativePng: any;

    constructor(buffer: Buffer) {
        if (!Buffer.isBuffer(buffer)) {
            throw new Error("Error decoding PNG. Input is not a buffer.");
        }
        this.nativePng = new __native_PngImage(buffer);
    }

    /**
     * The bit depth of the image.
     * Gathered from `png_get_bit_depth`.
     */
    public get bitDepth(): number { return this.nativePng.bitDepth; }

    /**
     * The amount of channels of the image.
     * Gathered from `png_get_channels`.
     */
    public get channels(): number { return this.nativePng.channels; }

    /**
     * The color type of the image as a string.
     * Gathered from `png_get_color_type`.
     */
    public get colorType(): ColorType { return this.nativePng.colorType; }

    /**
     * The width of the image.
     * Gathered from `png_get_image_height`.
     */
    public get height(): number { return this.nativePng.height; }

    /**
     * The width of the image.
     * Gathered from `png_get_image_width`.
     */
    public get width(): number { return this.nativePng.width; }

    /**
     * The interlace type of the image as a string, gathered from `png_get_interlace_type`.
     */
    public get interlaceType(): InterlaceType { return this.nativePng.interlaceType; }

    /**
     * The amount of bytes per row of the image.
     * Gathered from `png_get_rowbytes`.
     */
    public get rowBytes(): number { return this.nativePng.rowBytes; }

    /**
     * The horizontal offset of the image.
     * Gathered from `png_get_x_offset_pixels`.
     */
    public get offsetX(): number { return this.nativePng.offsetX; }

    /**
     * The vertical offset of the image.
     * Gathered from `png_get_y_offset_pixels`.
     */
    public get offsetY(): number { return this.nativePng.offsetY; }

    /**
     * The horizontal amount of pixels per meter of the image.
     * Gathered from `png_get_x_pixels_per_meter`.
     */
    public get pixelsPerMeterX(): number { return this.nativePng.pixelsPerMeterX; }

    /**
     * The vertical amount of pixels per meter of the image.
     * Gathered from `png_get_y_pixels_per_meter`.
     */
    public get pixelsPerMeterY(): number { return this.nativePng.pixelsPerMeterY; }

    /**
     * The buffer containing the data of the decoded image.
     */
    public get data(): Buffer { return this.nativePng.data; }

    /**
     * Will be `true` if the image's color type has an alpha channel and `false` otherwise.
     */
    public get alpha(): boolean {
        switch (this.colorType) {
            case ColorType.RGBA:
            case ColorType.GRAY_ALPHA:
                return true;
            default:
                return false;
        }
    }

    /**
     * Will encode this image to a PNG buffer.
     */
    public encode(): Buffer {
        const { alpha, width, height } = this;
        if (this.colorType !== ColorType.RGB && this.colorType !== ColorType.RGBA) {
            throw new Error("Can only encode images with RGB or RGBA color type.");
        }
        return encode(this.data, { width, height, alpha });
    }

    /**
     * Will encode this image and write it to the file at the specified path.
     *
     * @param path Path to the file to which the encoded PNG should be written.
     * @param callback An optional callback to use instead of the Promise API.
     *
     * @see writePngFile
     *
     * @return A Promise which resolves once the file is written or `undefined` if a callback was specified.
     */
    public write(path: string, callback?: WritePngFileCallback): Promise<void> | void {
        const { alpha, width, height } = this;
        if (this.colorType !== ColorType.RGB && this.colorType !== ColorType.RGBA) {
            throw new Error("Can only encode images with RGB or RGBA color type.");
        }
        return writePngFile(path, this.data, { width, height, alpha }, callback);
    }

    /**
     * Will encode this image and write it to the file at the specified path synchroneously.
     *
     * @param path Path to the file to which the encoded PNG should be written.
     *
     * @see writePngFileSync
     */
    public writeSync(path: string): void {
        const { alpha, width, height } = this;
        if (this.colorType !== ColorType.RGB && this.colorType !== ColorType.RGBA) {
            throw new Error("Can only encode images with RGB or RGBA color type.");
        }
        return writePngFileSync(path, this.data, { width, height, alpha });
    }
}
