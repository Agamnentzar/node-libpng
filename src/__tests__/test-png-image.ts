import { readFileSync } from "fs";
import { PngImage, ColorType } from "../png-image";
import { isColorRGB, isColorRGBA, isColorPalette, isColorGrayScale, isColorGrayScaleAlpha } from "../colors";
import { expectRedBlueGradient } from "./utils";

describe("PngImage", () => {
    it("reads the info of a normal png file", () => {
        const buffer = readFileSync(`${__dirname}/fixtures/red-blue-gradient-256px.png`);
        const image = new PngImage(buffer);

        expect(image.bitDepth).toBe(8);
        expect(image.channels).toBe(3);
        expect(image.colorType).toBe("rgb");
        expect(image.width).toBe(256);
        expect(image.height).toBe(256);
        expect(image.interlaceType).toBe("none");
        expect(image.rowBytes).toBe(256 * 3);
        expect(image.offsetX).toBe(0);
        expect(image.offsetY).toBe(0);
        expect(image.pixelsPerMeterX).toBe(0);
        expect(image.pixelsPerMeterY).toBe(0);
        expect(image.palette).toBeUndefined();
        expect(image.backgroundColor).toBeUndefined();
        expect(image.bytesPerPixel).toBe(3);
    });

    it("decodes a normal png file", () => {
        const inputBuffer = readFileSync(`${__dirname}/fixtures/red-blue-gradient-256px.png`);
        const { data } = new PngImage(inputBuffer);
        expectRedBlueGradient(data);
    });

    it("decodes a PNG file with ADAM7 interlacing", () => {
        const inputBuffer = readFileSync(`${__dirname}/fixtures/red-blue-gradient-256px-interlaced.png`);
        const { data } = new PngImage(inputBuffer);
        expectRedBlueGradient(data);
    });

    it("reads the palette of a ColorType.PALETTE image", () => {
        const inputBuffer = readFileSync(`${__dirname}/fixtures/indexed-16px.png`);
        const { palette } = new PngImage(inputBuffer);
        expect(palette.size).toBe(5);
        expect(Array.from(palette.entries())).toEqual([
            [0, [0, 0, 0]],
            [1, [128, 255, 64]],
            [2, [64, 128, 255]],
            [3, [255, 64, 128]],
            [4, [255, 128, 64]],
        ]);
    });

    it("reads the background color of an image with the background color specified", () => {
        const inputBuffer = readFileSync(`${__dirname}/fixtures/background-color.png`);
        const { backgroundColor } = new PngImage(inputBuffer);
        expect(backgroundColor).toEqual([255, 128, 64]);
    });

    it("converts XY coordinates to Index", () => {
        const inputBuffer = readFileSync(`${__dirname}/fixtures/orange-rectangle.png`);
        const image = new PngImage(inputBuffer);
        expect(image.width).toBe(32);
        expect(image.height).toBe(16);
        expect(image.bytesPerPixel).toBe(3);
        expect(image.toXY(5)).toEqual([1, 0]);
        expect(image.toXY(32 * 3 * 10 + 3 * 5)).toEqual([5, 10]);
        expect(image.toXY(image.data.length - 2)).toEqual([31, 15]);
    });

    it("converts Index to XY coordinates", () => {
        const inputBuffer = readFileSync(`${__dirname}/fixtures/orange-rectangle.png`);
        const image = new PngImage(inputBuffer);
        expect(image.width).toBe(32);
        expect(image.height).toBe(16);
        expect(image.bytesPerPixel).toBe(3);
        expect(image.toIndex(1, 0)).toEqual(3);
        expect(image.toIndex(5, 10)).toEqual(32 * 3 * 10 + 3 * 5);
        expect(image.toIndex(31, 15)).toEqual(image.data.length - 3);
    });

    it("converts Index to XY coordinates", () => {
        const inputBuffer = readFileSync(`${__dirname}/fixtures/orange-rectangle.png`);
        const image = new PngImage(inputBuffer);
        expect(image.width).toBe(32);
        expect(image.height).toBe(16);
        expect(image.bytesPerPixel).toBe(3);
        expect(image.toIndex(1, 0)).toEqual(3);
        expect(image.toIndex(5, 10)).toEqual(32 * 3 * 10 + 3 * 5);
        expect(image.toIndex(31, 15)).toEqual(image.data.length - 3);
    });

    [
        {
            colorType: ColorType.GRAY_SCALE,
            file: `${__dirname}/fixtures/grayscale-gradient-16px.png`,
            checker: isColorGrayScale,
        },
        {
            colorType: ColorType.GRAY_SCALE_ALPHA,
            file: `${__dirname}/fixtures/grayscale-alpha-gradient-16px.png`,
            checker: isColorGrayScaleAlpha,
        },
        {
            colorType: ColorType.RGB,
            file: `${__dirname}/fixtures/orange-rectangle.png`,
            checker: isColorRGB,
        },
        {
            colorType: ColorType.RGBA,
            file: `${__dirname}/fixtures/opaque-rectangle.png`,
            checker: isColorRGBA,
        },
        {
            colorType: ColorType.PALETTE,
            file: `${__dirname}/fixtures/indexed-16px.png`,
            checker: isColorPalette,
        },
    ].forEach(({ colorType, file, checker }) => {
        it(`detects the correct color with an image of color type ${colorType}`, () => {
            const inputBuffer = readFileSync(file);
            const image = new PngImage(inputBuffer);
            expect(image.colorType).toBe(colorType);
            const color = image.at(10, 10);
            expect(checker(color)).toBe(true);
            expect(color).toMatchSnapshot();
        });
    });

    describe("export methods", () => {
        const somePngImage = new PngImage(readFileSync(`${__dirname}/fixtures/orange-rectangle.png`));

        describe("encode", () => {
            it("encodes a PNG file with the colortype being RGB", () => {
                expect(somePngImage.encode().toString("hex")).toMatchSnapshot();
            });
        });

        describe("write", () => {
            it("writes a PNG file with the colortype being RGB", async () => {
                const path = `${__dirname}/../../tmp-png-image-write.png`;
                await somePngImage.write(path);
                const fromDisk = readFileSync(path);
                expect(fromDisk.toString("hex")).toMatchSnapshot();
            });
        });

        describe("writeSync", () => {
            it("writes a PNG file with the colortype being RGB", () => {
                const path = `${__dirname}/../../tmp-png-image-write-sync.png`;
                somePngImage.writeSync(path);
                const fromDisk = readFileSync(path);
                expect(fromDisk.toString("hex")).toMatchSnapshot();
            });
        });
    });
});
