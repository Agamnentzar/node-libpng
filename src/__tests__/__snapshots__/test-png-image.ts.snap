// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PngImage detects the correct color with an image of color type gray-scale 1`] = `
Array [
  85,
]
`;

exports[`PngImage detects the correct color with an image of color type gray-scale-alpha 1`] = `
Array [
  85,
  127,
]
`;

exports[`PngImage detects the correct color with an image of color type palette 1`] = `
Array [
  3,
]
`;

exports[`PngImage detects the correct color with an image of color type rgb 1`] = `
Array [
  255,
  128,
  64,
]
`;

exports[`PngImage detects the correct color with an image of color type rgba 1`] = `
Array [
  255,
  128,
  64,
  127,
]
`;

exports[`PngImage export methods encode encodes a PNG file with the colortype being RGB 1`] = `"89504e470d0a1a0a0000000d4948445200000020000000100802000000f862ea0e000000204944415438cb63fcdfe0c0404bc0c44063306ac1a805a3168c5a306a01350000fad701df0311050e0000000049454e44ae426082"`;

exports[`PngImage export methods write writes a PNG file with the colortype being RGB 1`] = `"89504e470d0a1a0a0000000d4948445200000020000000100802000000f862ea0e000000204944415438cb63fcdfe0c0404bc0c44063306ac1a805a3168c5a306a01350000fad701df0311050e0000000049454e44ae426082"`;

exports[`PngImage export methods writeSync writes a PNG file with the colortype being RGB 1`] = `"89504e470d0a1a0a0000000d4948445200000020000000100802000000f862ea0e000000204944415438cb63fcdfe0c0404bc0c44063306ac1a805a3168c5a306a01350000fad701df0311050e0000000049454e44ae426082"`;
